src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -248,13 +248,14 @@ private bool TryValidatePseudoHeaders()
             // enabling the use of HTTP to interact with non - HTTP services.
             // A common example is TLS termination.
             var headerScheme = HttpRequestHeaders.HeaderScheme.ToString();
+            HttpRequestHeaders.HeaderScheme = default; // Suppress pseduo headers from the public headers collection.
```

Comment:
With this change, we'd be setting the pseudo-header values onto HttpRequestHeaders, then later clearing them. What about just never setting these values onto HttpRequestHeaders at all? We have control over where these values are set in the `OnHeader` methods in HTTP/2 and HTTP/3.

Reply:
Hmmm, but the headers collection does provide some nice benefits like automatic string reuse.

Reply:
> What about just never setting these values onto HttpRequestHeaders at all?

That was the alternative but there are several trade-offs:
- OnHeaders has multiple overloads, we'd have to check for the pseudo headers in each of them on every header.
- It would be a more significant and risky change for rc2.
- The value re-use feature.
- Sharing the code between HTTP/2 and 3
- Handling weird corner cases like duplicate headers.

HttpRequestHeaders is already well optimized for these. Accessing and removing the values later has comparatively little cost due to the direct properties. The main reason to move these out of header collection is if we were concerned about the size. That's a design decision we can revisit in the future.